### ISSUE TITLE
Adding fiscal_start month option in the case of `ceiling_date(x, 'quarter')`?

### ISSUE THREAD

**Issue by tomcardoso** [tidyverse/lubridate#947]
Hi there,

Big fan of lubridate, it's served me well over the past few years. I've something of an edge case feature request.

Right now, `ceiling_date(x, 'quarter')` returns starting quarter dates with the assumption quarters begin on Jan. 1. Would it be possible to add an extra parameter, similar to the `quarter` function, for a month value for `fiscal_start`? The math when dealing with fiscal years not beginning with Jan. 1 get complicated really fast, especially if you have a date that is, say, in 2021 but actually in the 2020 or 2022 fiscal year due to how the quarters break down.

In the case of a fiscal year starting on Aug. 1, for instance, we'd be looking at:
```
Q1: 2020-08-01
Q2: 2020-11-01
Q3: 2021-02-01
Q4: 2021-05-01
```

So if this were to be implemented, here's what we'd expect to see:

```r
dates <- ymd(c('1980-01-01', '1983-04-19', '2001-11-15', '2006-09-23'))
ceiling_date(x, 'quarter', fiscal_start = 8)
[1] "1979-11-01" "1983-02-01" "2001-11-01" "2006-08-01"
```
Alternatively, if `ceiling_date` could recognize the `quarter(x, with_year = TRUE, fiscal_start = 8)` outputs, could pass them that way.

Would this be a worthy inclusion? I'd certainly end up using it a lot – doing quarters math for any company that doesn't have FY starting on Jan. 1 (which is most of them) is always a pain as a result.

**Comment by tomcardoso** [comment]
Or, thinking about this more, perhaps this should all be contained in the `quarter` API itself. Perhaps some usage like this?
```r
quarter(x, as_date = TRUE, fiscal_start = 8)
```

The code would then look something like this, I think (still need to double-check I didn't forget to account for some edge cases…):
```r
dates_as_quarters <- function(x, with_year = FALSE, as_date = TRUE, fiscal_start = 1) {
  fs <- (fiscal_start - 1) %% 12
  shifted <- seq(fs, 11 + fs) %% 12 + 1
  m <- month(x)
  quarters <- rep(1:4, each = 3)
  s <- match(m, shifted)
  q <- quarters[s]
  # new section below to handle our new dates
  if (as_date) {
    starting_months <- shifted[seq(1, length(shifted), 3)]
    final_years <- year(x) - (starting_months[q] > m)
    ymd(paste(final_years, starting_months[q], '01', sep = '-'))
  } else if (with_year) {
    this_year <- if (fs != 0) (fs + 1):12
    quarter_value <- year(x) + (m %in% this_year) + q/10
  } else {
    q
  }
}
```

This should net:
```r
dates <- ymd(c('1980-01-01', '1983-04-19', '2001-11-15', '2006-09-23'))
dates_as_quarters(dates, as_date = TRUE, fiscal_start = 8)
# "1979-11-01" "1983-02-01" "2001-11-01" "2006-08-01"
```

**Comment by vspinu** [comment]
I think adding `fiscal_start` to round_date is wranted, but it's probably not a trivial task. As you say, the logic is complicated. If you can open a PR I would be happy to accept though. 

`as_date` to quater is not bad, but the UI of that function becomes rather ugly. I would rather prefer a "type" or "return" argument that would take "quarter", "full" and "date" options. But it's too late to change. 

How about new function `quarter2date` or `as_date` method? We would need to add an atribute "fiscal_start" and class to the output of `quarter` function.

**Comment by tomcardoso** [comment]
A new function makes sense to me. If we're thinking of function design, it might make sense to also add the option to go to the beginning or end of a quarter. So perhaps something like:
```r
quarter_as_date(x, fiscal_start = 1, type = 'end') # For, say, "2021-02-14", would return "2021-03-31"
quarter_as_date(x, fiscal_start = 1, type = 'start') # For, say, "2021-02-14", would return "2021-01-01"
```
I'm guessing we'd probably want to take date objects for that function, and not something like `2021.2` or whatever, right?

As for `ceiling_date`, I'm happy to write a PR. I'm not super familiar with how tight tidyverse maintainers expect the code to be, but happy to give it the old college try.

**Comment by tomcardoso** [comment]
Hey @vspinu, just to follow up… what do you think of the above? I can take a crack at writing something next week if you think that approach makes sense.

**Comment by vspinu** [comment]
HI @tomcardoso Sorry for coming later on this. 

I thought a bit about this. I think going along your original proposal is the best UI. Let's add new argument "type" to quarter and set  `with_year = type == "quarter" with the explicit statement int he docs that `with_year` is for compatibility only and is soft-deprecated. WDYT?

How about the following types: `quarter`, `quarter_year`, `date_first`, `date_last`?

**Comment by tomcardoso** [comment]
I think that makes a lot of sense. So the API would be something like:

```r
dates <- ymd(c("1980-01-01", "1983-04-19", "2001-11-15", "2006-09-23", "2015-07-04"))

quarter(dates, type = "quarter", fiscal_start = 8)
[1] 2 3 2 1 4

quarter(dates, type = "quarter_year", fiscal_start = 8)
[1] 1979.2 1982.3 2001.2 2006.1 2014.4

quarter(dates, type = "date_first", fiscal_start = 8)
[1] "1979-11-01" "1983-02-01" "2001-11-01" "2006-08-01" "2015-05-01"

quarter(dates, type = "date_last", fiscal_start = 8)
[1] "1979-01-31" "1983-04-30" "2002-01-31" "2006-10-31" "2006-07-31"
```

Does that look right? Should be a minimal amount of work to implement, and we could keep the default behaviour to `type = "quarter"`.

**Comment by tomcardoso** [comment]
We could *probably* simplify `date_first` and `date_last` to simply `date`, but since people often want both I figure it's not a bad idea to save them the effort of having to do something like `quarter(dates, type = "date", fiscal_start = 8) + months(3) - days(1)`.

**Comment by vspinu** [comment]
We can also have "date" and "date_last"  as types.

**Comment by tomcardoso** [comment]
Got it. Here's what I'm thinking. If it looks good, I can put together some tests and the new code into a PR.

```r
quarter <- function(x, type = 'quarter', with_year = type == 'quarter', fiscal_start = 1) {
  fs <- (fiscal_start - 1) %% 12
  shifted <- seq(fs, 11 + fs) %% 12 + 1
  m <- month(x)
  quarters <- rep(1:4, each = 3)
  s <- match(m, shifted)
  q <- quarters[s]

  # new section below to handle new types
  if (type == 'quarter') {
    q
  } else if (type == 'quarter_year') {
    first_quarter_month <- ifelse(fs != 0, fs + 1, 0)
    year(x) - (m < first_quarter_month) + (q / 10)
  } else if (type %in% c('date_first', 'date_last')) {
    starting_months <- shifted[seq(1, length(shifted), 3)]
    final_years <- year(x) - (starting_months[q] > m)
    quarter_starting_dates <- ymd(paste(final_years, starting_months[q], '01', sep = '-'))
    if (type == 'date_first') {
      quarter_starting_dates
    } else if (type == 'date_last') {
      quarter_starting_dates + month(3) - days(1)
    }
  }
}
```

**Comment by vspinu** [comment]
> I can put together some tests and the new code into a PR.

Yes please. Let's discuss on the PR. 

Instead of cascading if-else it's better to use `switch`. Instead of `+months(3) - days(1)` please use `add_with_rollback`. Instead of `ymd(paste ..` `make_date`

Also please move `with_year` as last argument and allow for `quarter` to be logical, in which case it's equivalent to specifying with_year. This is necessary to avoid breaking existing positional calls. Otherwise we would need to keep quarter lat parameter.

**Comment by vspinu** [comment]
> Should we perhaps name the type param something else? Is it a reserved term in R?

Not a reserved symbol. It can be rettype, or return, or kind. I slightly prefer `rettype` and `type` but `type` is shorter.

**Comment by tomcardoso** [comment]
Sounds good, I'll pull something together next week. One question tough:

> Also please move with_year as last argument and allow for quarter to be logical, in which case it's equivalent to specifying with_year. 

What do you mean by this? Handle cases where `fiscal_start` could be logical? Or `type`?

**Comment by vspinu** [comment]
> Handle cases where fiscal_start could be logical? Or type?

Arh, sorry. I meant `type` argument should be logical in order to accomodate existing positional calls with `fiscal_start` in the second argument.

**Comment by tomcardoso** [comment]
I'll submit a PR soon. It'll include a few comments here and there for things we can discuss. One question before I do, though. One of the tests has this:
```r
posct <- ymd_hms("2010-11-03 13:45:59")
expect_equal(quarter(posct, with_year = TRUE, fiscal_start = 11), 2011.1)
```

But since that quarter has a `fiscal_start` of 11, should it technically return `2010.1` (which is what the new code does) as it's the first quarter for that new year? Or is the `lubridate` style to return the numeric year during which the FY **ends**, in which case it'd be 2011.1 (as the last quarter would end on 2011-10-31)?

**Comment by vspinu** [comment]
All tests should be consistent.  The end of the year is the reference I think. The expected behavior is that seen in excel. See https://github.com/tidyverse/lubridate/issues/890 .

**Comment by tomcardoso** [comment]
Thanks for that – I've fixed the code so that it behaves as it used to (using 12th month year in the quarter-year notation), and the PR has now been submitted.

**Comment by DavisVaughan** [comment]
Closed by #955
